fontscan/fontmap.go

```diff
@@ -0,0 +1,394 @@
+//
+//	// set the font description
+//	fontMap.SetQuery(Query{Families: []string{"Arial", "serif"}}) // regular Aspect
+//
+//	// `fontMap` is now ready for text shaping, using the `ResolveFace` method
+//
+// Note that [FontMap] is NOT safe for concurrent use, but several font maps may coexist
+// in an application.
+//
+// [FontMap] is designed to work with an index built by scanning the system fonts,
+// which is a costly operation (see [UseSystemFonts] for more details).
+// A lightweight alternative is provided by the [FindFont] function, which only uses
+// file paths to select a font.
+type FontMap struct {
+	// cache of already loaded faces
+	faces map[Location]font.Face
+
+	// the database to query, either loaded from an index
+	// or populated with the UseSystemFonts and AddFont method
+	database fontSet
+
+	// the candidates for the current query, which influences ResolveFace output
+	candidates candidates
+
+	// internal buffers used in SetQuery
+	footprintsBuffer scoredFootprints
+	cribleBuffer     familyCrible
+
+	query Query // current query
+
+	// cached value of the last footprint index
+	// selected by ResolveFace
+	lastFootprintIndex int
+}
+
+// NewFontMap return a new font map,
+// which should be filled with the `UseSystemFonts`
+// or `AddFont` methods.
+func NewFontMap() *FontMap {
+	return &FontMap{
+		faces:              make(map[Location]font.Face),
+		cribleBuffer:       make(familyCrible),
+		lastFootprintIndex: -1,
+	}
+}
+
+// UseSystemFonts loads the system fonts and adds them to the font map.
+// This method is safe for concurrent use, but should only be called once
+// per font map.
+// The first call of this method trigger a rather long scan.
+// A per-application on-disk cache is used to speed up subsequent initialisations.
+func (fm *FontMap) UseSystemFonts() error {
+	// safe for concurrent use; subsequent calls are no-ops
+	err := initSystemFonts()
+	if err != nil {
+		return err
+	}
+
+	// systemFonts is read-only, so may be used concurrently
+	fm.database = append(fm.database, systemFonts.flatten()...)
+
+	fm.buildCandidates()
+
+	return nil
+}
+
+// systemFonts is a global index of the system fonts.
+// initSystemFontsOnce protects the initial assignment,
+// and `systemFonts` use is then read-only
+var (
+	systemFonts         systemFontsIndex
+	initSystemFontsOnce sync.Once
+)
+
+// initSystemFonts scan the system fonts and update `SystemFonts`.
+// If the returned error is nil, `SystemFonts` is guaranteed to contain
+// at least one valid font.Face.
+// It is protected by sync.Once, and is then safe to use by multiple goroutines.
+func initSystemFonts() error {
+	var err error
+
+	initSystemFontsOnce.Do(func() {
+		const cacheFile = "font_index.cache"
+
+		// load an existing index
+		var execPath string
+		execPath, err = os.Executable()
+		if err != nil {
+			err = fmt.Errorf("resolving index cache path: %s", err)
+			return
+		}
+
+		cachePath := filepath.Join(filepath.Dir(execPath), cacheFile)
```

Comment:
This isn't a safe way to determine where to store the cache. In macOS application bundles, for example, the directory containing the executables is readonly. On Linux, it would be highly unusual to find `font_index.cache` in a `bin` directory, and some application sandboxing may prevent this outright.

Every OS has an application-writable configuration directory though. On Linux it's usually `$XDG_CONFIG_HOME/application`, on macOS it's `/Users/me/Library/Application Support/application`, and on Windows it's `C:\User\me\AppData\Local\application`. Android and iOS are compatible (I think) with the Linux and macOS strategies, which only leaves WASM as the tricky case. There's probably a way to use local storage for this in WASM.

I think we discussed this in the past as well, but if we were able to make the font cache format backwards and forwards compatible, it would be possible to share the cache across applications. Imagine Fyne and Gio applications being able to leverage the same system font cache on the same computer, rather than each generating their own! Then we could place it somewhere generic (like a folder called `gotext`) instead of in a per-application location.

Reply:
Fair enough, thank you !
I've extracted the cache directory in `cacheDir` and used `os.UserConfigDir` as first implementation. 
Feel free to update it :) In particular, I'm not sure this works on mobile...
For WASM, we would need `syscall/js` right ? like in [this library](https://github.com/jonhteper/wasm/blob/main/storage.go) ?

Reply:
Yeah, we'd need `syscall/js` for sure. However, I don't know how we would go about discovering available fonts. Seems like it'd be a big effort. We could initially just make the system font scan return an empty set when running in the browswer. Then we could expand after someone has time to investigate how to do the operation correctly within that runtime environment.

Reply:
> Android and iOS are compatible (I think) with the Linux and macOS strategies, which only leaves WASM as the tricky case. There's probably a way to use local storage for this in WASM.

I don't think that is accurate for iOS - the cache location will be in the app sandbox not in the `$USER/Library/...` location won't it?

> I think we discussed this in the past as well, but if we were able to make the font cache format backwards and forwards compatible, it would be possible to share the cache across applications.

We can't do that for mobile due to app sandbox issues I think.

> We could initially just make the system font scan return an empty set when running in the browswer.

I'd support this on a first pass. System lookups from a web browser is likely a big topic!

Reply:
@whereswaldon Sure thing, thank you for digging this up. Perhaps you could edit `DefaultFontDirectories` in `scan.go` accordingly ?

Reply:
I'll be happy to. I've done a local rebase of this atop latest `main`, which I plan to put up as a separate draft PR so you can double-check my work. I'll include this change there.